src/nvim/eval/typval.c

```diff
@@ -1205,9 +1205,11 @@ void tv_dict_watcher_notify(dict_T *const dict, const char *const key,
     DictWatcher *watcher = tv_dict_watcher_node_data(w);
     if (!watcher->busy && tv_dict_watcher_matches(watcher, key)) {
       rettv = TV_INITIAL_VALUE;
+      dict->dv_refcount++;
       watcher->busy = true;
       callback_call(&watcher->callback, 3, argv, &rettv);
       watcher->busy = false;
+      dict->dv_refcount--;
```

Comment:
Should maybe use `tv_dict_unref`?  https://github.com/neovim/neovim/blob/1a6efa4719c906fe67f2c21ec1a2b33ae92ca17b/src/nvim/eval/typval.c#L1419-L1424